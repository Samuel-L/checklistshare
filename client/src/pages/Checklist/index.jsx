import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import { fetchChecklist } from '../../redux-modules/checklist-fetcher';

export class Checklist extends Component {
  componentDidMount() {
    let url = this.props.location.pathname;
    url = url.substring(1, url.length); // remove '/' from start of url
    this.props.fetchChecklist(url);
  }

  render() {
    return (
      <div>
        !error
          show checklist
        else
          show checklist not found
      </div>
    );
  }
}

Checklist.propTypes = {
  location: PropTypes.shape({
    pathname: PropTypes.string.isRequired,
  }),
};

Checklist.defaultProps = {
  location: {
    pathname: '',
  },
};

const mapStateToProps = state => ({
  fetching: state.checklistFetcherReducer.fetching,
  fetched: state.checklistFetcherReducer.fetched,
  error: state.checklistFetcherReducer.error,
  checklist: state.checklistFetcherReducer.checklist,
});

const mapDispatchToProps = dispatch => ({
  fetchChecklist: url => dispatch(fetchChecklist(url)),
});

Checklist.propTypes = {
  fetchChecklist: PropTypes.func.isRequired,
};

export default connect(mapStateToProps, mapDispatchToProps)(Checklist);
