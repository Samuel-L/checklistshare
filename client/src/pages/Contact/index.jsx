import React from 'react';
import PropTypes from 'prop-types';

import { withStyles } from 'material-ui/styles';
import Grid from 'material-ui/Grid';
import Typography from 'material-ui/Typography';

const styles = theme => ({
  base: {},
  title: {
    marginTop: '20px',
  },
  text: {
    marginTop: '10px',
  },
  link: {
    color: theme.palette.primary.A700,
    textDecoration: 'none',
    '&:hover': {
      color: theme.palette.primary.dark,
    },
  },
});

const Contact = props => (
  <Grid container>
    <Grid item xs={12}>
      <Typography
        className={props.classes.title}
        variant="display1"
        align="center"
        gutterBottom
      >
        Contact
      </Typography>
    </Grid>
    <Grid item xs={12}>
      <Typography className={props.classes.text}>
        Feedback and bug findings can be sent to samuell.person@gmail.com.
      </Typography>
      <Typography className={props.classes.text}>
        Source code can be found <a className={props.classes.link} href="https://github.com/Samuel-L/checklistshare">here</a>.
      </Typography>
    </Grid>
  </Grid>
);

Contact.propTypes = {
  classes: PropTypes.objectOf(PropTypes.string).isRequired,
};

export default withStyles(styles)(Contact);
