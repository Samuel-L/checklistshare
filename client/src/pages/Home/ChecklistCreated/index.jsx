import React, { Component } from 'react';
import PropTypes from 'prop-types';

import Grid from 'material-ui/Grid';
import Typography from 'material-ui/Typography';
import Button from 'material-ui/Button';
import { CircularProgress } from 'material-ui/Progress';
import { CopyToClipboard } from 'react-copy-to-clipboard';
import Snackbar from 'material-ui/Snackbar';

import clientBaseURL from '../../../utils/client-helpers';

export class SuccessfulCreation extends Component {
  state = { copied: false };

  render() {
    const url = `${clientBaseURL}/${this.props.url}`;

    return (
      <React.Fragment>
        <Grid item xs={12} md={6}>
          <Typography
            variant="headline"
            align="center"
            gutterBottom
          >
            Checklist created!
          </Typography>
          <Typography
            variant="body2"
            align="center"
          >
            {url}
          </Typography>
          <CopyToClipboard text={url} onCopy={() => this.setState({ copied: true })} style={{ float: 'right' }}>
            <Button variant="raised" color="primary">
              Copy url
            </Button>
          </CopyToClipboard>
        </Grid>
        <Snackbar
          anchorOrigin={{
            vertical: 'top',
            horizontal: 'left',
          }}
          open={this.state.copied}
          message="URL was copied successfully!"
        />
      </React.Fragment>
    );
  }
}

SuccessfulCreation.propTypes = {
  url: PropTypes.string.isRequired,
};

const FailedCreation = () => (
  <Grid item>
    <Typography
      variant="headline"
      align="center"
      gutterBottom
    >
      Something went wrong on our end. Please try again.
    </Typography>
  </Grid>
);

const ChecklistCreated = props => (
  <Grid container justify="center">
    { props.adding && !props.added
      ?
        <Grid item>
          <CircularProgress size={70} />
        </Grid>
      :
        [
          (props.added && props.url ? <SuccessfulCreation key="0" url={props.url} /> : <FailedCreation key="0" />),
        ]
    }
  </Grid>
);

ChecklistCreated.propTypes = {
  adding: PropTypes.bool,
  added: PropTypes.bool,
  url: PropTypes.string,
};

ChecklistCreated.defaultProps = {
  adding: false,
  added: false,
  url: '',
};

export default ChecklistCreated;
