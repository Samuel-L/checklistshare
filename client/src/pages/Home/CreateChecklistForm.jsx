import React from 'react';
import PropTypes from 'prop-types';

import { withStyles } from 'material-ui/styles';
import TextField from 'material-ui/TextField';
import Button from 'material-ui/Button';

import ChecklistItem from './ChecklistItem';

const styles = theme => ({
  textField: {
    margin: '5px',
  },
  buttonGroup: {
    [theme.breakpoints.up('md')]: {
      width: '150px',
      float: 'right',
    },
    [theme.breakpoints.down('sm')]: {
      position: 'absolute',
      bottom: '0',
      width: '100%',
    },
  },
  buttonDiv: {
    width: '100%',
  },
  button: {
    width: '100%',
    margin: '1px',
    color: 'white',
    backgroundColor: theme.palette.primary.A700,
    '&:hover': {
      backgroundColor: theme.palette.primary.A700,
      color: 'white',
    },
  },
});

const CreateChecklistForm = props => (
  <form
    id="create-list-form"
    onSubmit={props.handleSubmit}
  >
    <TextField
      required
      id="name"
      label="Title"
      value={props.title}
      onChange={props.handleTextFieldChange}
      helperText="Please give your checklist a title"
      fullWidth
      margin="normal"
      className={props.classes.textField}
    />
    {
      props.items.map((item, idx) => (
        <ChecklistItem
          key={item.id}
          item={item}
          idx={idx}
          handleItemTextChange={() => {}}
          handleItemDelete={() => {}}
        />
      ))
    }
    <div className={props.classes.buttonGroup}>
      <div className={props.classes.buttonDiv}>
        <Button
          id="add-item-button"
          className={props.classes.button}
          variant="flat"
          color="primary"
          onClick={props.handleAddItem}
        >
          Add Item
        </Button>
      </div>
      <div className={props.classes.buttonDiv}>
        <Button
          className={props.classes.button}
          variant="flat"
          color="primary"
        >
          Create List
        </Button>
      </div>
    </div>
  </form>
);

CreateChecklistForm.propTypes = {
  classes: PropTypes.objectOf(PropTypes.string).isRequired,
  handleSubmit: PropTypes.func.isRequired,
  handleTextFieldChange: PropTypes.func.isRequired,
  handleAddItem: PropTypes.func.isRequired,
  title: PropTypes.string.isRequired,
  items: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number,
    name: PropTypes.string,
  })).isRequired,
};

CreateChecklistForm.defaultProps = {};

export default withStyles(styles)(CreateChecklistForm);
