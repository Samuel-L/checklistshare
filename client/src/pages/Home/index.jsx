import React, { Component } from 'react';

import Grid from 'material-ui/Grid';
import Typography from 'material-ui/Typography';

import CreateChecklistForm from './CreateChecklistForm';

export class Home extends Component {
  state = {
    items: [{ id: 0, name: '' }],
  };

  handleAddItem = () => {
    const { items } = this.state;
    this.setState({
      items: this.state.items.concat([{ id: items.length, name: '' }]),
    });
  };

  handleDeleteItem = (idx) => {
    this.setState({
      items: this.state.items.filter((item, iidx) => idx !== iidx),
    }, () => {
      // Ensure that the id field is unique
      const newItems = this.state.items.map((item, iidx) => ({
        ...item, id: iidx,
      }));

      this.setState({ items: newItems });
    });
  };

  render() {
    return (
      <Grid container justify="center" alignItems="center">
        <Grid item xs={12}>
          <Typography variant="display1" align="center">
            Create checklist
          </Typography>
        </Grid>
        <Grid item xs={12} lg={6}>
          <CreateChecklistForm
            handleSubmit={() => {}}
            handleTextFieldChange={() => {}}
            handleAddItem={this.handleAddItem}
            handleDeleteItem={this.handleDeleteItem}
            title="Checklist title"
            items={this.state.items}
          />
        </Grid>
      </Grid>
    );
  }
}

export default Home;
