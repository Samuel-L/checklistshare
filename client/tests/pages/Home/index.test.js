import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';

import { Home } from '../../../src/pages/Home';

describe('pages: Home', () => {
  describe('rendering', () => {
    const tree = renderer.create(<Home addChecklist={jest.fn()} />).toJSON();

    it('matches snapshot', () => {
      expect(tree).toMatchSnapshot(); 
    });
  });

  describe('functionality', () => {
    let wrapper;

    beforeEach(() => {
      wrapper = mount(<Home addChecklist={jest.fn()} />);
    });

    it('handles changing the checklist title', () => {
      wrapper.find('[id="title"]').at(3).simulate('change',
        { target: { name: 'title', value: 'Checklist title' } }
      );

      expect(wrapper.state('title')).toBe('Checklist title');
    });

    it('handles changing item text', () => {
      expect(wrapper.state('items')[0].name).toBe('');
      wrapper.find('[id="item-0"]').at(3).simulate('change',
        { target: { name: 'item-0', value: 'Item text' } }
      );
      expect(wrapper.state('items')[0].name).toBe('Item text');
    });

    it('handles adding new items', () => {
      expect(wrapper.state('items').length).toEqual(1); 
      wrapper.find('#add-item-button').at(0).simulate('click');
      expect(wrapper.state('items').length).toEqual(2);
    });

    it('handles deleting items', () => {
      expect(wrapper.state('items').length).toEqual(1);
      wrapper.find('#delete-item-button').at(0).simulate('click');
      expect(wrapper.state('items').length).toEqual(0);
    });

    it('handles creating an entire checklist', () => {
      const finishedChecklistState = {
        title: 'Checklist title',
        items: [
          { id: 0, name: 'Item name 1' },
          { id: 1, name: 'Item name 2'},
          { id: 2, name: 'Item name 3'}
        ],
        submitConfirmationModalOpen: true,
      };

      wrapper.find('[id="title"]').at(3).simulate('change',
        { target: { name: 'title', value: 'Checklist title' } }
      );

      wrapper.find('#delete-item-button').at(0).simulate('click');
      
      wrapper.find('#add-item-button').at(0).simulate('click');
      wrapper.find('[id="item-0"]').at(3).simulate('change',
        { target: { name: 'item-0', value: 'Item name 1' } }
      );

      wrapper.find('#add-item-button').at(0).simulate('click');
      wrapper.find('[id="item-1"]').at(3).simulate('change',
        { target: { name: 'item-1', value: 'Item name 2' } }
      );

      wrapper.find('#add-item-button').at(0).simulate('click');
      wrapper.find('[id="item-2"]').at(3).simulate('change',
        { target: { name: 'item-2', value: 'Item name 3' } }
      );

      wrapper.find('#create-list-button').at(0).simulate('click');

      expect(wrapper.state()).toEqual(finishedChecklistState);
    });
  });
});
