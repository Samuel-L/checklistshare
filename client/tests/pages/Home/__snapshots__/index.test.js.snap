// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages: Home should render 1`] = `
<div
  className="MuiGrid-container-1 MuiGrid-align-items-xs-center-9 MuiGrid-justify-xs-center-18"
>
  <div
    className="MuiGrid-item-2 MuiGrid-grid-xs-12-38"
  >
    <h1
      className="MuiTypography-root-91 MuiTypography-display1-95 MuiTypography-alignCenter-104"
    >
      Create checklist
    </h1>
  </div>
  <div
    className="MuiGrid-item-2 MuiGrid-grid-xs-12-38 MuiGrid-grid-lg-6-71"
  >
    <form
      id="create-list-form"
      onSubmit={[Function]}
    >
      <div
        aria-describedby="name-helper-text"
        className="MuiFormControl-root-119 MuiFormControl-marginNormal-120 MuiFormControl-fullWidth-122 CreateChecklistForm-textField-115"
        onBlur={[Function]}
        onFocus={[Function]}
      >
        <label
          className="MuiFormLabel-root-128 MuiInputLabel-root-123 MuiInputLabel-formControl-124 MuiInputLabel-animated-127 MuiInputLabel-shrink-126"
          data-shrink={true}
          htmlFor="name"
        >
          Title
          <span
            className="MuiFormLabel-asterisk-132"
          >
             *
          </span>
        </label>
        <div
          className="MuiInput-root-133 MuiInput-fullWidth-140 MuiInput-formControl-134 MuiInput-underline-137"
        >
          <input
            aria-invalid={false}
            aria-required={true}
            autoComplete={undefined}
            autoFocus={undefined}
            className="MuiInput-input-141"
            defaultValue={undefined}
            disabled={false}
            id="name"
            name={undefined}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyDown={undefined}
            onKeyUp={undefined}
            placeholder={undefined}
            readOnly={undefined}
            required={true}
            rows={undefined}
            type="text"
            value="Checklist title"
          />
        </div>
        <p
          className="MuiFormHelperText-root-146"
          id="name-helper-text"
        >
          Please give your checklist a title
        </p>
      </div>
      <div
        className="ChecklistItem-root-150"
      >
        <div
          aria-describedby={undefined}
          className="MuiFormControl-root-119 MuiFormControl-marginNormal-120 ChecklistItem-textField-151"
          onBlur={[Function]}
          onFocus={[Function]}
        >
          <label
            className="MuiFormLabel-root-128 MuiInputLabel-root-123 MuiInputLabel-formControl-124 MuiInputLabel-animated-127"
            data-shrink={false}
            htmlFor="item-0"
          >
            Item
            <span
              className="MuiFormLabel-asterisk-132"
            >
               *
            </span>
          </label>
          <div
            className="MuiInput-root-133 MuiInput-formControl-134 MuiInput-underline-137"
          >
            <input
              aria-invalid={false}
              aria-required={true}
              autoComplete={undefined}
              autoFocus={undefined}
              className="MuiInput-input-141"
              defaultValue={undefined}
              disabled={false}
              id="item-0"
              name={undefined}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onKeyDown={undefined}
              onKeyUp={undefined}
              placeholder={undefined}
              readOnly={undefined}
              required={true}
              rows={undefined}
              type="text"
              value=""
            />
          </div>
        </div>
        <button
          aria-label="delete"
          className="MuiButtonBase-root-168 MuiButton-root-153 MuiButton-raised-158 ChecklistItem-icon-152"
          disabled={false}
          id="delete-item-button"
          onBlur={[Function]}
          onClick={undefined}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex="0"
          type="button"
        >
          <span
            className="MuiButton-label-154"
          >
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root-171"
              color={undefined}
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
              />
            </svg>
          </span>
          <span
            className="MuiTouchRipple-root-177"
          />
        </button>
      </div>
      <div
        className="CreateChecklistForm-buttonGroup-116"
      >
        <div
          className="CreateChecklistForm-buttonDiv-117"
        >
          <button
            className="MuiButtonBase-root-168 MuiButton-root-153 MuiButton-flatPrimary-155 CreateChecklistForm-button-118"
            disabled={false}
            onBlur={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex="0"
            type="button"
          >
            <span
              className="MuiButton-label-154"
            >
              Add Item
            </span>
            <span
              className="MuiTouchRipple-root-177"
            />
          </button>
        </div>
        <div
          className="CreateChecklistForm-buttonDiv-117"
        >
          <button
            className="MuiButtonBase-root-168 MuiButton-root-153 MuiButton-flatPrimary-155 CreateChecklistForm-button-118"
            disabled={false}
            onBlur={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex="0"
            type="button"
          >
            <span
              className="MuiButton-label-154"
            >
              Create List
            </span>
            <span
              className="MuiTouchRipple-root-177"
            />
          </button>
        </div>
      </div>
    </form>
  </div>
</div>
`;
